### PR TITLE
chore(lib/crypto): consolation of crypto / crypto module

### PR DESCRIPTION
**Pull request by jbenden**
Consolidate usage of RC4, MD5, SHA-1, SHA-256, and HMAC.

Add support for OpenSSL >= 3.0 and/or OpenSSL with/without
support for the various algorithms listed above.

### CODE REVIEW

**Comment by aircrack-ng** [comment]
Could you add an indication of what openssl version is being used in the `./configure` summary?

**Comment by jbenden** [comment]
Oddly, not that easy - especially when mixing cross-compilation...

**Comment by Mister-X-** [comment]
It's a nice to have, so don't worry too much about it if it's not that easy.

In any case, performance testing doesn't show much of a difference in loss of performance. On a system with AVX2, here are the numbers for ubuntu 20.04 docker with openssl v1 vs the same with openssl v3 (running on an Ubuntu 22.04 in KVM with 2 cores):

|         |  openssl v1  |  openssl v3  |
| ---------|--------------|--------------|
| ba8a93b aka HEAD  |    8250 k/s  |              |
| PR #2326  |    8284 k/s  |    8250 k/s  |

Testing on the same base system but using an Ubuntu 22.04 docker (which has openssl v3), we get around 8012 k/s

File: test/cryptounittest/test-calc-pmk.c [hunk 1]
```diff
@@ -33,7 +33,7 @@
 #include "aircrack-ng/crypto/crypto.h"
 #include "tests.h"
 
-#define PLEN 40
+#define PLEN 32 //40
```
Comment (Mister-X-) [inline test/cryptounittest/test-calc-pmk.c:36]: Why is there a change of data length (cutting off 8 last bytes)?

File: build/m4/aircrack_ng_crypto.m4 [hunk 2]
```diff
@@ -1,6 +1,6 @@
 dnl Aircrack-ng
 dnl
-dnl Copyright (C) 2017 Joseph Benden <joe@benden.us>
+dnl Copyright (C) 2022 Joseph Benden <joe@benden.us>
```
Comment (Mister-X-) [inline build/m4/aircrack_ng_crypto.m4]: Shouldn't this be "2017,2022" ?

**Comment by Mister-X-** [comment]
Running `aircrack-ng -S` segfaults after a few runs (3 or 4). Using the following `Dockerfile` (in case that matters, running on Ubuntu 22.04 host):

```
FROM ubuntu:22.04
RUN apt-get update && \
	apt upgrade -y && \
	DEBIAN_FRONTEND=noninteractive apt install -y --no-install-recommends tzdata && \
	apt install git build-essential autoconf automake libtool pkg-config libnl-3-dev libnl-genl-3-dev libssl-dev ethtool shtool rfkill zlib1g-dev libpcap-dev libsqlite3-dev libpcre3-dev libhwloc-dev libcmocka-dev hostapd wpasupplicant tcpdump screen iw usbutils -y && \
	rm -rf /var/lib/dpkg/status-old /var/lib/apt/lists/*
WORKDIR /aircrack-ng
RUN git clone https://github.com/aircrack-ng/aircrack-ng pr2326 && cd pr2326 && git pull origin pull/2326/head && git fetch && autoreconf -i && ./configure --with-experimental && make && make install
```

**Comment by Mister-X-** [comment]
`make check` is successful on MIPS Big Endian, no failed test.

File: test/cryptounittest/test-calc-pmk.c (hunk 1 above)
Comment (jbenden) [reply test/cryptounittest/test-calc-pmk.c:36]: The length does not match what is used throughout the rest of the code; so I need to know why it was marked 40 to begin with! :)

File: build/m4/aircrack_ng_crypto.m4 (hunk 2 above)
Comment (jbenden) [reply build/m4/aircrack_ng_crypto.m4]: Good argument for it!

**Comment by jbenden** [comment]
Would you mind capturing a few backtraces from that Ubuntu container?

**Comment by Mister-X-** [comment]
[backtraces.tar.gz](https://github.com/aircrack-ng/aircrack-ng/files/9620926/backtraces.tar.gz)

**Comment by jbenden** [comment]
Could I talk you into decoding those?

**Comment by Mister-X-** [comment]
I'm assuming they were missing the debug information. Here you go, fresh ones with debug info for GDB: [coredump.zip](https://github.com/aircrack-ng/aircrack-ng/files/9688039/coredump.zip)

**Comment by Mister-X-** [comment]
This results in at best 2 keys/sec (on average 1 key/sec) on an emulated MIPS BE when running `aircrack-ng -S`, whereas #2335 (which I assume will do the same as ba8a93beef861eaabca8cdafc1b3edec19c915f2) does 60 keys/sec on the exact same system.

Note: It takes an hour to compile.

**Comment by jbenden** [comment]
Not much can be easily done when upstream (openssl project) refuses to publicly support low-level API interfaces.

In the long-term, we could provide all cryptography function; and use no external library. We will be forced to anyway, as the upstream vendors deprecate old algorithms...

So for now, do we like this PR? :sweat_smile:

**Comment by Mister-X-** [comment]
Yes, aside from the crashes of benchmark mode.

**Comment by Mister-X-** [comment]
Maybe we can create a small library for these cryptographic functions. Maybe other tools would make use of it.

**Comment by Mister-X-** [comment]
> Not much can be easily done when upstream (openssl project) refuses to publicly support low-level API interfaces.
> ...

One thing I forgot to mention is that MIPS BE system is Debian 10, so it was with OpenSSL 1.x

**Comment by jbenden** [comment]
@Mister-X- I cannot reproduce the seg. faults, using my slightly modified file. I cannot get your file to work without modification. I've included my modified file content below:

``` Dockerfile
FROM ubuntu:22.04
RUN apt-get update && \
        apt upgrade -y && \
        DEBIAN_FRONTEND=noninteractive apt install -y --no-install-recommends tzdata && \
        apt install git build-essential autoconf automake libtool pkg-config libnl-3-dev libnl-genl-3-dev libssl-dev ethtool shtool rfkill zlib1g-dev libpcap-dev libsqlite3-dev libpcre3-dev libhwloc-dev libcmocka-dev hostapd wpasupplicant tcpdump screen iw usbutils -y && \
        rm -rf /var/lib/dpkg/status-old /var/lib/apt/lists/*
WORKDIR /aircrack-ng
RUN set -x \
 && git config --global user.name 'Aircrack-ng Support' \
 && git config --global user.email 'support@aircrack-ng.org' \
 && git clone https://github.com/aircrack-ng/aircrack-ng pr2326 \
 && cd pr2326 \
 && git pull --rebase origin pull/2326/head \
 && git fetch \
 && autoreconf -if \
 && ./configure --with-experimental \
 && err=0; while [ $err -eq 0 ]; do make -j4 check || err=1; make clean; done
```
PS: **The above script will INFINITELY BUILD - YOU"VE BEEN WARNED!**

**Comment by Mister-X-** [comment]
It's not when building that it crashes, it's when running `aircrack-ng -S`.

What do you mean you can't get my file to work without modification?

**Comment by jbenden** [comment]
Yes, but my test-case ran just fine for about an hour. So I don't see a problem...

Your script throws build errors... Here:
```
Step 4/4 : RUN git clone https://github.com/aircrack-ng/aircrack-ng pr2326 && cd pr2326 && git pull origin pull/2326/head && git fetch && autoreconf -i && ./configure --with-experimental && make && make install
 ---> Running in 83119bab4d0d
Cloning into 'pr2326'...
From https://github.com/aircrack-ng/aircrack-ng
 * branch              refs/pull/2326/head -> FETCH_HEAD
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint:
hint:   git config pull.rebase false  # merge (the default strategy)
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint:
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.
The command '/bin/sh -c git clone https://github.com/aircrack-ng/aircrack-ng pr2326 && cd pr2326 && git pull origin pull/2326/head && git fetch && autoreconf -i && ./configure --with-experimental && make && make install' returned a non-zero code: 128
```
Just use my `Dockerfile`! :)

**Comment by jbenden** [comment]
@Mister-X- Um, I did find a crash... Researching...

**Comment by jbenden** [comment]
The crash was caused by a race condition during shutdown, from a call to `exit()` that circumvented shutting down worker threads.

Additionally, a few FreeBSD 13 caught problems were resolved.

There is an odd bug with a macOS runner currently, not sure what is going on. I thought that md5/md5sum was missing, but installing via Homebrew did not resolve the problem.

Please re-review.

@Mister-X-

**Comment by Mister-X-** [comment]
Now works fine, no crashes. We can take care of the Mac runner later on. All good.

**Comment by jbenden** [comment]
Closes: #2230
Closes: #2231
Closes: #2222

**Pull request merged**